File: .config/anto426/osd/osd.py
#!/usr/bin/env python3
"""Centred GTK OSD slider for volume, mic and brightness."""

from __future__ import annotations

import os
import signal
import sys
from pathlib import Path

import gi

gi.require_version("Gtk", "3.0")
gi.require_version("Gdk", "3.0")
try:
    gi.require_version("GtkLayerShell", "0.1")
    from gi.repository import GtkLayerShell  # noqa: E402

    HAS_LAYER_SHELL = True
except (ImportError, ValueError):
    GtkLayerShell = None
    HAS_LAYER_SHELL = False
from gi.repository import Gdk, GLib, Gtk  # noqa: E402

STATE_FILE = Path(
    os.environ.get(
        "ANTO426_OSD_STATE",
        f"{os.environ.get('XDG_RUNTIME_DIR', '/tmp')}/anto426-osd.state",
    )
)
PID_FILE = Path(
    os.environ.get(
        "ANTO426_OSD_PID",
        f"{os.environ.get('XDG_RUNTIME_DIR', '/tmp')}/anto426-osd.pid",
    )
)
CSS_FILE = Path(__file__).with_name("style.css")
COLORS_FILE = Path(
    os.environ.get("XDG_CONFIG_HOME", str(Path.home() / ".config"))
) / "colors" / "colors.css"
HIDE_MS = 1400
OSD_WIDTH = 340
OSD_HEIGHT = 84
BAR_WIDTH = 280
BOTTOM_MARGIN = 48

FALLBACK_CSS = b"""
@define-color background #1e1e2e;
@define-color surface #313244;
@define-color foreground #f6f7fb;
@define-color muted #b9c4d2;
@define-color accent #8cb8e4;
@define-color border #6c7086;
@define-color panel-bg rgba(30, 30, 46, 0.62);
@define-color overlay-bg rgba(30, 30, 46, 0.28);
@define-color item-bg rgba(49, 50, 68, 0.18);
@define-color item-bg-active rgba(140, 184, 228, 0.42);
@define-color border-medium rgba(108, 112, 134, 0.34);
@define-color background-alpha @panel-bg;
@define-color surface-alpha @item-bg;
"""

ICONS = {
    "volume": "󰕾",
    "brightness": "󰃠",
    "mic": "󰍬",
}

TITLES = {
    "volume": "Volume",
    "brightness": "Luminosità",
    "mic": "Microfono",
}


class OsdWindow(Gtk.Window):
    def __init__(self) -> None:
        super().__init__(type=Gtk.WindowType.TOPLEVEL)
        self.set_decorated(False)
        self.set_resizable(False)
        self.set_keep_above(True)
        self.set_skip_taskbar_hint(True)
        self.set_skip_pager_hint(True)
        self.set_accept_focus(False)
        self.set_type_hint(Gdk.WindowTypeHint.NOTIFICATION)
        self.set_app_paintable(True)
        self.get_style_context().add_class("osd-root")
        self.set_name("anto426-osd")
        try:
            self.set_wmclass("anto426-osd", "anto426-osd")
        except AttributeError:
            pass
        self._configure_surface()

        self._load_css()

        outer = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=0)
        outer.get_style_context().add_class("osd-box")
        self.add(outer)

        header = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=0)
        outer.pack_start(header, False, False, 0)

        self.icon_label = Gtk.Label()
        self.icon_label.get_style_context().add_class("osd-icon")
        header.pack_start(self.icon_label, False, False, 0)

        text_col = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=0)
        header.pack_start(text_col, True, True, 0)

        self.title_label = Gtk.Label(xalign=0)
        self.title_label.get_style_context().add_class("osd-title")
        text_col.pack_start(self.title_label, False, False, 0)

        self.value_label = Gtk.Label(xalign=0)
        self.value_label.get_style_context().add_class("osd-value")
        text_col.pack_start(self.value_label, False, False, 0)

        self.progress = Gtk.ProgressBar()
        self.progress.set_show_text(False)
        self.progress.set_size_request(BAR_WIDTH, 8)
        self.progress.get_style_context().add_class("osd-track")
        outer.pack_start(self.progress, False, False, 0)

        self._hide_id: int | None = None
        self.show_all()
        self._centre()

    def _load_css(self) -> None:
        screen = Gdk.Screen.get_default()
        if screen is None:
            return

        fallback = Gtk.CssProvider()
        fallback.load_from_data(FALLBACK_CSS)
        Gtk.StyleContext.add_provider_for_screen(
            screen,
            fallback,
            Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION,
        )

        if COLORS_FILE.exists():
            colors = Gtk.CssProvider()
            colors.load_from_path(str(COLORS_FILE))
            Gtk.StyleContext.add_provider_for_screen(
                screen,
                colors,
                Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION + 1,
            )

        provider = Gtk.CssProvider()
        provider.load_from_path(str(CSS_FILE))
        Gtk.StyleContext.add_provider_for_screen(
            screen,
            provider,
            Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION + 2,
        )

    def _configure_surface(self) -> None:
        screen = self.get_screen()
        visual = screen.get_rgba_visual() if screen is not None else None
        if visual is not None:
            self.set_visual(visual)

        if HAS_LAYER_SHELL and GtkLayerShell is not None:
            GtkLayerShell.init_for_window(self)
            GtkLayerShell.set_namespace(self, "anto426-osd")
            GtkLayerShell.set_layer(self, GtkLayerShell.Layer.OVERLAY)
            GtkLayerShell.set_keyboard_mode(self, GtkLayerShell.KeyboardMode.NONE)
            GtkLayerShell.set_exclusive_zone(self, -1)
            GtkLayerShell.set_anchor(self, GtkLayerShell.Edge.LEFT, True)
            GtkLayerShell.set_anchor(self, GtkLayerShell.Edge.BOTTOM, True)
            return

        self.set_position(Gtk.WindowPosition.CENTER)

    def _centre(self) -> None:
        display = Gdk.Display.get_default()
        monitor = display.get_primary_monitor() if display else None
        if monitor is None and display is not None:
            monitor = display.get_monitor(0)
        if monitor is None:
            return
        geo = monitor.get_geometry()
        x = max(0, (geo.width - OSD_WIDTH) // 2)
        if HAS_LAYER_SHELL and GtkLayerShell is not None:
            GtkLayerShell.set_margin(self, GtkLayerShell.Edge.LEFT, x)
            GtkLayerShell.set_margin(self, GtkLayerShell.Edge.BOTTOM, BOTTOM_MARGIN)
            return

        y = max(0, geo.height - OSD_HEIGHT - BOTTOM_MARGIN)
        window = self.get_window()
        if window is not None:
            window.move(geo.x + x, geo.y + y)

    def _schedule_hide(self) -> None:
        if self._hide_id is not None:
            GLib.source_remove(self._hide_id)
        self._hide_id = GLib.timeout_add(HIDE_MS, self._hide)

    def _hide(self) -> bool:
        Gtk.main_quit()
        return False

    def update(self, kind: str, value: int, muted: bool = False) -> None:
        kind = kind if kind in ICONS else "volume"
        value = max(0, min(100, int(value)))
        muted = bool(muted)

        self.icon_label.set_text("󰝟" if muted else ICONS[kind])
        self.title_label.set_text(TITLES[kind])
        self.value_label.set_text("Muto" if muted else f"{value}%")

        self.progress.set_fraction(0.0 if muted else value / 100)
        if muted:
            self.progress.get_style_context().add_class("muted")
        else:
            self.progress.get_style_context().remove_class("muted")

        if not self.get_visible():
            self.show_all()
        self._centre()
        self._schedule_hide()


def read_state() -> tuple[str, int, bool]:
    try:
        raw = STATE_FILE.read_text(encoding="utf-8").strip().split()
    except OSError:
        raw = []
    if len(raw) < 2:
        return "volume", 0, False
    kind = raw[0]
    try:
        value = int(float(raw[1]))
    except ValueError:
        value = 0
    muted = len(raw) > 2 and raw[2] in {"1", "true", "yes", "muted"}
    return kind, value, muted


def write_state(kind: str, value: int, muted: bool = False) -> None:
    STATE_FILE.parent.mkdir(parents=True, exist_ok=True)
    STATE_FILE.write_text(
        f"{kind} {value} {1 if muted else 0}\n",
        encoding="utf-8",
    )


def cleanup_pid_file() -> None:
    try:
        current = PID_FILE.read_text(encoding="utf-8").strip()
    except OSError:
        return
    if current == str(os.getpid()):
        try:
            PID_FILE.unlink()
        except OSError:
            pass


def main() -> int:
    if len(sys.argv) >= 2 and sys.argv[1] != "daemon":
        kind = sys.argv[1]
        value = int(sys.argv[2]) if len(sys.argv) > 2 else 0
        muted = len(sys.argv) > 3 and sys.argv[3] in {"1", "true", "muted"}
        write_state(kind, value, muted)

    kind, value, muted = read_state()
    win = OsdWindow()
    win.update(kind, value, muted)

    def refresh_from_state() -> bool:
        k, v, m = read_state()
        win.update(k, v, m)
        return False

    def on_usr1(*_args: object) -> None:
        GLib.idle_add(refresh_from_state)

    signal.signal(signal.SIGUSR1, on_usr1)
    try:
        Gtk.main()
        return 0
    finally:
        cleanup_pid_file()


if __name__ == "__main__":
    raise SystemExit(main())

File: .config/anto426/osd/osd.c
#include <gtk/gtk.h>
#include <gtk-layer-shell.h>
#include <glib-unix.h>
#include <signal.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#define OSD_WIDTH 380
#define OSD_HEIGHT 64
#define BAR_WIDTH 260
#define BOTTOM_MARGIN 48
#define HIDE_MS 1400

static char state_file_path[1024];
static char pid_file_path[1024];
static char colors_file_path[1024];
static char style_file_path[1024];

static char state_kind[32] = "volume";
static int state_value = 0;
static int state_muted = 0;

static GtkWidget *icon_label = NULL;
static GtkWidget *title_label = NULL;
static GtkWidget *value_label = NULL;
static GtkWidget *progress_bar = NULL;
static guint hide_timeout_id = 0;

static const char *FALLBACK_CSS = 
    "@define-color background #1e1e2e;\n"
    "@define-color surface #313244;\n"
    "@define-color foreground #f6f7fb;\n"
    "@define-color muted #b9c4d2;\n"
    "@define-color accent #8cb8e4;\n"
    "@define-color border #6c7086;\n"
    "@define-color panel-bg rgba(30, 30, 46, 0.62);\n"
    "@define-color overlay-bg rgba(30, 30, 46, 0.28);\n"
    "@define-color item-bg rgba(49, 50, 68, 0.18);\n"
    "@define-color item-bg-active rgba(140, 184, 228, 0.42);\n"
    "@define-color border-medium rgba(108, 112, 134, 0.34);\n"
    "@define-color background-alpha @panel-bg;\n"
    "@define-color surface-alpha @item-bg;\n";

static void resolve_paths(void) {
    // 1. STATE_FILE
    const char *env_state = getenv("ANTO426_OSD_STATE");
    if (env_state) {
        strncpy(state_file_path, env_state, sizeof(state_file_path) - 1);
    } else {
        const char *runtime_dir = getenv("XDG_RUNTIME_DIR");
        if (runtime_dir) {
            snprintf(state_file_path, sizeof(state_file_path), "%s/anto426-osd.state", runtime_dir);
        } else {
            strcpy(state_file_path, "/tmp/anto426-osd.state");
        }
    }

    // 2. PID_FILE
    const char *env_pid = getenv("ANTO426_OSD_PID");
    if (env_pid) {
        strncpy(pid_file_path, env_pid, sizeof(pid_file_path) - 1);
    } else {
        const char *runtime_dir = getenv("XDG_RUNTIME_DIR");
        if (runtime_dir) {
            snprintf(pid_file_path, sizeof(pid_file_path), "%s/anto426-osd.pid", runtime_dir);
        } else {
            strcpy(pid_file_path, "/tmp/anto426-osd.pid");
        }
    }

    // 3. COLORS_FILE
    const char *config_home = getenv("XDG_CONFIG_HOME");
    if (config_home) {
        snprintf(colors_file_path, sizeof(colors_file_path), "%s/colors/colors.css", config_home);
    } else {
        const char *home = getenv("HOME");
        if (home) {
            snprintf(colors_file_path, sizeof(colors_file_path), "%s/.config/colors/colors.css", home);
        } else {
            strcpy(colors_file_path, "/tmp/colors.css");
        }
    }

    // 4. CSS_FILE (style.css in the same directory as the binary)
    char exe_path[1024];
    ssize_t len = readlink("/proc/self/exe", exe_path, sizeof(exe_path) - 1);
    if (len != -1) {
        exe_path[len] = '\0';
        char *dir = strrchr(exe_path, '/');
        if (dir) {
            *dir = '\0';
            snprintf(style_file_path, sizeof(style_file_path), "%s/style.css", exe_path);
        } else {
            strcpy(style_file_path, "style.css");
        }
    } else {
        strcpy(style_file_path, "style.css");
    }
}

static void read_state(void) {
    FILE *f = fopen(state_file_path, "r");
    if (!f) return;
    if (fscanf(f, "%31s %d %d", state_kind, &state_value, &state_muted) < 2) {
        strcpy(state_kind, "volume");
        state_value = 0;
        state_muted = 0;
    }
    fclose(f);
}

static void write_state(const char *kind, int value, int muted) {
    FILE *f = fopen(state_file_path, "w");
    if (f) {
        fprintf(f, "%s %d %d\n", kind, value, muted);
        fclose(f);
    }
}

static void write_pid_file(void) {
    FILE *f = fopen(pid_file_path, "w");
    if (f) {
        fprintf(f, "%d\n", getpid());
        fclose(f);
    }
}

static void cleanup_pid_file(void) {
    FILE *f = fopen(pid_file_path, "r");
    if (f) {
        int read_pid = 0;
        if (fscanf(f, "%d", &read_pid) == 1) {
            if (read_pid == getpid()) {
                unlink(pid_file_path);
            }
        }
        fclose(f);
    }
}

static void configure_surface(GtkWidget *window) {
    GdkScreen *screen = gtk_window_get_screen(GTK_WINDOW(window));
    GdkVisual *visual = gdk_screen_get_rgba_visual(screen);
    if (visual) {
        gtk_widget_set_visual(window, visual);
    }

    if (gtk_layer_is_supported()) {
        gtk_layer_init_for_window(GTK_WINDOW(window));
        gtk_layer_set_namespace(GTK_WINDOW(window), "anto426-osd");
        gtk_layer_set_layer(GTK_WINDOW(window), GTK_LAYER_SHELL_LAYER_OVERLAY);
        gtk_layer_set_keyboard_mode(GTK_WINDOW(window), GTK_LAYER_SHELL_KEYBOARD_MODE_NONE);
        gtk_layer_set_exclusive_zone(GTK_WINDOW(window), -1);
        gtk_layer_set_anchor(GTK_WINDOW(window), GTK_LAYER_SHELL_EDGE_LEFT, FALSE);
        gtk_layer_set_anchor(GTK_WINDOW(window), GTK_LAYER_SHELL_EDGE_RIGHT, FALSE);
        gtk_layer_set_anchor(GTK_WINDOW(window), GTK_LAYER_SHELL_EDGE_TOP, FALSE);
        gtk_layer_set_anchor(GTK_WINDOW(window), GTK_LAYER_SHELL_EDGE_BOTTOM, TRUE);
    } else {
        g_warning("GtkLayerShell is NOT supported. Running in fallback X11 mode.");
    }
}

static void center_window(GtkWidget *window) {
    if (gtk_layer_is_supported()) {
        gtk_layer_set_margin(GTK_WINDOW(window), GTK_LAYER_SHELL_EDGE_BOTTOM, BOTTOM_MARGIN);
    } else {
        GdkDisplay *display = gdk_display_get_default();
        GdkMonitor *monitor = NULL;
        if (display) {
            monitor = gdk_display_get_primary_monitor(display);
            if (!monitor) {
                monitor = gdk_display_get_monitor(display, 0);
            }
        }
        if (!monitor) return;

        GdkRectangle geo;
        gdk_monitor_get_geometry(monitor, &geo);

        int x = (geo.width - OSD_WIDTH) / 2;
        if (x < 0) x = 0;
        int y = geo.height - OSD_HEIGHT - BOTTOM_MARGIN;
        gtk_window_move(GTK_WINDOW(window), geo.x + x, geo.y + y);
    }
}

static void load_css(GtkWidget *window) {
    GdkScreen *screen = gtk_window_get_screen(GTK_WINDOW(window));
    if (!screen) return;

    GtkCssProvider *fallback_provider = gtk_css_provider_new();
    gtk_css_provider_load_from_data(fallback_provider, FALLBACK_CSS, -1, NULL);
    gtk_style_context_add_provider_for_screen(
        screen,
        GTK_STYLE_PROVIDER(fallback_provider),
        GTK_STYLE_PROVIDER_PRIORITY_APPLICATION
    );
    g_object_unref(fallback_provider);

    if (access(colors_file_path, F_OK) == 0) {
        GtkCssProvider *colors_provider = gtk_css_provider_new();
        gtk_css_provider_load_from_path(colors_provider, colors_file_path, NULL);
        gtk_style_context_add_provider_for_screen(
            screen,
            GTK_STYLE_PROVIDER(colors_provider),
            GTK_STYLE_PROVIDER_PRIORITY_APPLICATION + 1
        );
        g_object_unref(colors_provider);
    }

    if (access(style_file_path, F_OK) == 0) {
        GtkCssProvider *style_provider = gtk_css_provider_new();
        gtk_css_provider_load_from_path(style_provider, style_file_path, NULL);
        gtk_style_context_add_provider_for_screen(
            screen,
            GTK_STYLE_PROVIDER(style_provider),
            GTK_STYLE_PROVIDER_PRIORITY_APPLICATION + 2
        );
        g_object_unref(style_provider);
    }
}

static gboolean on_hide_timeout(gpointer data) {
    (void)data;
    gtk_main_quit();
    hide_timeout_id = 0;
    return FALSE;
}

static void schedule_hide(void) {
    if (hide_timeout_id > 0) {
        g_source_remove(hide_timeout_id);
    }
    hide_timeout_id = g_timeout_add(HIDE_MS, on_hide_timeout, NULL);
}

static void update_osd(GtkWidget *window) {
    read_state();

    const char *icon_str = "";
    const char *title_str = "";
    static char value_str[32];

    if (strcmp(state_kind, "brightness") == 0) {
        icon_str = "󰃠";
        title_str = "Luminosità";
    } else if (strcmp(state_kind, "mic") == 0) {
        icon_str = state_muted ? "󰝟" : "󰍬";
        title_str = "Microfono";
    } else { // volume
        icon_str = state_muted ? "󰝟" : "󰕾";
        title_str = "Volume";
    }

    if (state_muted) {
        strcpy(value_str, "Muto");
    } else {
        snprintf(value_str, sizeof(value_str), "%d%%", state_value);
    }

    gtk_label_set_text(GTK_LABEL(icon_label), icon_str);
    char *title_upper = g_utf8_strup(title_str, -1);
    gtk_label_set_text(GTK_LABEL(title_label), title_upper);
    g_free(title_upper);
    gtk_label_set_text(GTK_LABEL(value_label), value_str);

    double fraction = state_muted ? 0.0 : ((double)state_value / 100.0);
    gtk_progress_bar_set_fraction(GTK_PROGRESS_BAR(progress_bar), fraction);

    GtkStyleContext *context = gtk_widget_get_style_context(progress_bar);
    if (state_muted) {
        gtk_style_context_add_class(context, "muted");
    } else {
        gtk_style_context_remove_class(context, "muted");
    }

    if (!gtk_widget_get_visible(window)) {
        gtk_widget_show_all(window);
    }
    center_window(window);
    schedule_hide();
}

static gboolean on_sigusr1(gpointer user_data) {
    GtkWidget *window = GTK_WIDGET(user_data);
    update_osd(window);
    return TRUE;
}

static gboolean on_draw(GtkWidget *widget, cairo_t *cr, gpointer user_data) {
    (void)widget;
    (void)user_data;
    cairo_set_source_rgba(cr, 0.0, 0.0, 0.0, 0.0);
    cairo_set_operator(cr, CAIRO_OPERATOR_SOURCE);
    cairo_paint(cr);
    return FALSE;
}

int main(int argc, char *argv[]) {
    resolve_paths();

    if (argc >= 2 && strcmp(argv[1], "daemon") != 0) {
        const char *kind = argv[1];
        int value = (argc > 2) ? atoi(argv[2]) : 0;
        int muted = 0;
        if (argc > 3 && (strcmp(argv[3], "1") == 0 || strcmp(argv[3], "true") == 0 || strcmp(argv[3], "muted") == 0)) {
            muted = 1;
        }
        write_state(kind, value, muted);
    }

    gtk_init(&argc, &argv);

    GtkWidget *window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_decorated(GTK_WINDOW(window), FALSE);
    gtk_window_set_resizable(GTK_WINDOW(window), FALSE);
    gtk_window_set_keep_above(GTK_WINDOW(window), TRUE);
    gtk_window_set_skip_taskbar_hint(GTK_WINDOW(window), TRUE);
    gtk_window_set_skip_pager_hint(GTK_WINDOW(window), TRUE);
    gtk_window_set_accept_focus(GTK_WINDOW(window), FALSE);
    gtk_window_set_type_hint(GTK_WINDOW(window), GDK_WINDOW_TYPE_HINT_NOTIFICATION);
    gtk_widget_set_name(window, "anto426-osd");
    gtk_style_context_add_class(gtk_widget_get_style_context(window), "osd-root");

    gtk_widget_set_app_paintable(window, TRUE);
    g_signal_connect(window, "draw", G_CALLBACK(on_draw), NULL);

    gtk_widget_set_size_request(window, OSD_WIDTH, OSD_HEIGHT);

    configure_surface(window);
    load_css(window);

    GtkWidget *outer = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 0);
    gtk_widget_set_name(outer, "osd-box");
    gtk_widget_set_size_request(outer, OSD_WIDTH, OSD_HEIGHT);
    gtk_container_add(GTK_CONTAINER(window), outer);

    icon_label = gtk_label_new("");
    gtk_style_context_add_class(gtk_widget_get_style_context(icon_label), "osd-icon");
    gtk_box_pack_start(GTK_BOX(outer), icon_label, FALSE, FALSE, 0);

    GtkWidget *center_box = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
    gtk_widget_set_name(center_box, "osd-center-box");
    gtk_box_pack_start(GTK_BOX(outer), center_box, TRUE, TRUE, 0);

    title_label = gtk_label_new("");
    gtk_widget_set_halign(title_label, GTK_ALIGN_START);
    gtk_style_context_add_class(gtk_widget_get_style_context(title_label), "osd-title");
    gtk_box_pack_start(GTK_BOX(center_box), title_label, FALSE, FALSE, 0);

    progress_bar = gtk_progress_bar_new();
    gtk_progress_bar_set_show_text(GTK_PROGRESS_BAR(progress_bar), FALSE);
    gtk_widget_set_size_request(progress_bar, BAR_WIDTH, 8);
    gtk_style_context_add_class(gtk_widget_get_style_context(progress_bar), "osd-track");
    gtk_box_pack_start(GTK_BOX(center_box), progress_bar, FALSE, FALSE, 0);

    value_label = gtk_label_new("");
    gtk_widget_set_halign(value_label, GTK_ALIGN_END);
    gtk_label_set_xalign(GTK_LABEL(value_label), 1.0);
    gtk_style_context_add_class(gtk_widget_get_style_context(value_label), "osd-value");
    gtk_box_pack_start(GTK_BOX(outer), value_label, FALSE, FALSE, 0);

    write_pid_file();
    update_osd(window);

    g_unix_signal_add(SIGUSR1, on_sigusr1, window);

    gtk_main();

    cleanup_pid_file();
    return 0;
}
